src/mame/drivers/dfs500.cpp

```diff
@@ -0,0 +1,1036 @@
+		case 1: input_bitmap = &m_input2->get_bitmap(); break;
+		case 2: input_bitmap = &m_input3->get_bitmap(); break;
+		case 3: input_bitmap = &m_input4->get_bitmap(); break;
+	}
+
+	if (input_bitmap)
+	{
+		// convert arbitrary sized ARGB32 image to a full-screen image
+		double stepx = (double)input_bitmap->width() / VIDEO_WIDTH;
+		double stepy = (double)input_bitmap->height() / VIDEO_HEIGHT;
+
+		for (unsigned screen_y = screen.visible_area().min_y; screen_y <= screen.visible_area().max_y; screen_y++)
+		{
+			for (unsigned screen_x = screen.visible_area().min_x; screen_x <= screen.visible_area().max_x; screen_x++)
+			{
+				bitmap.pix(screen_y, screen_x) = input_bitmap->pix((int)((double)(screen_y % VIDEO_HEIGHT) * stepy),
+				                                                   (int)((double)(screen_x % VIDEO_WIDTH) * stepx));
+			}
+		}
+	}
+	return 0;
+}
+
+uint8_t dfs500_state::pit_r(offs_t offset)
+{
+	// CXQ71054P (Programmable Timer / Counter)
+	return m_pit->read((offset>>1) & 3); // addressed by CPU's address bits AB2 and AB1
+}
+
+void dfs500_state::pit_w(offs_t offset, uint8_t data)
+{
+	// CXQ71054P (Programmable Timer / Counter)
+	m_pit->write((offset>>1) & 3, data); // addressed by CPU's address bits AB2 and AB1
+}
+
+uint8_t dfs500_state::cpanel_pit_r(offs_t offset)
+{
+	// CXQ71054P (Programmable Timer / Counter)
+	return m_cpanel_pit->read((offset>>1) & 3); // addressed by CPU's address bits AB2 and AB1
+}
+
+void dfs500_state::cpanel_pit_w(offs_t offset, uint8_t data)
+{
+	// CXQ71054P (Programmable Timer / Counter)
+	m_cpanel_pit->write((offset>>1) & 3, data); // addressed by CPU's address bits AB2 and AB1
+}
+
+uint8_t dfs500_state::pic_r(offs_t offset)
+{
+	// PD71059C (Programmable Interrupt Controller)
+	return m_pic->read((offset>>1) & 1); // addressed by CPU's address bit AB1
+}
+
+void dfs500_state::pic_w(offs_t offset, uint8_t data)
+{
+	// PD71059C (Programmable Interrupt Controller)
+	m_pic->write((offset>>1) & 1, data); // addressed by CPU's address bit AB1
+}
+
+void dfs500_state::rombank1_entry_w(offs_t offset, uint8_t data)
+{
+	m_rombank1->set_entry(((data >> 1) & 0x40) | (data & 0x3f));
+}
+
+void dfs500_state::rombank2_entry_w(offs_t offset, uint8_t data)
+{
+	m_rombank2->set_entry(((data >> 1) & 0x40) | (data & 0x3f));
+}
+
+void dfs500_state::input_select_w(offs_t offset, uint8_t data)
+{
+	// Selects sources of video input on the AD-76 board.
+	m_input_sel_A = (data >> 3) & 0x7;
+	m_input_sel_B = data & 0x7;
+}
+
+void dfs500_state::WA0_w(offs_t offset, uint16_t data)
+{
+	m_effectcpu_latch16 = data;
+	m_TOC = true;
+}
+
+uint16_t dfs500_state::RA0_r(offs_t offset)
+{
+	return m_maincpu_latch16;
+	m_TOE = false;
+}
+
+uint16_t dfs500_state::RA1_r(offs_t offset)
+{
+	// "TEST, 1, OPT2, OPT1, RFLD, VD, TOC, TOE"
+	uint8_t value = 0x40;
+	// FIXME! Add other signals.
+	if (m_TOC) value |= 0x02;
+	if (m_TOE) value |= 0x01;
+	value |= ((ioport("DSW_S3")->read() & 0x0f) << 8); // (Unknown)
```

Comment:
Please use `required_ioport`. Do not use string lookup-based ioport access at runtime.

Reply:
done